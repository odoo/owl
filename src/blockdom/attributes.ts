import type { Setter } from "./block_compiler";

const { setAttribute: elemSetAttribute, removeAttribute } = Element.prototype;
const tokenList = DOMTokenList.prototype;
const tokenListAdd = tokenList.add;
const tokenListRemove = tokenList.remove;
const isArray = Array.isArray;
const { split, trim } = String.prototype;
const wordRegexp = /\s+/;

/**
 * We regroup here all code related to updating attributes in a very loose sense:
 * attributes, properties and classs are all managed by the functions in this
 * file.
 */

function setAttribute(this: HTMLElement, key: string, value: any) {
  switch (value) {
    case false:
    case undefined:
      removeAttribute.call(this, key);
      break;
    case true:
      elemSetAttribute.call(this, key, "");
      break;
    default:
      elemSetAttribute.call(this, key, value);
  }
}

export function createAttrUpdater(attr: string): Setter<HTMLElement> {
  return function (this: HTMLElement, value: any) {
    setAttribute.call(this, attr, value);
  };
}

export function attrsSetter(this: HTMLElement, attrs: any) {
  if (isArray(attrs)) {
    setAttribute.call(this, attrs[0], attrs[1]);
  } else {
    for (let k in attrs) {
      setAttribute.call(this, k, attrs[k]);
    }
  }
}

export function attrsUpdater(this: HTMLElement, attrs: any, oldAttrs: any) {
  if (isArray(attrs)) {
    const name = attrs[0];
    const val = attrs[1];
    if (name === oldAttrs[0]) {
      if (val === oldAttrs[1]) {
        return;
      }
      setAttribute.call(this, name, val);
    } else {
      removeAttribute.call(this, oldAttrs[0]);
      setAttribute.call(this, name, val);
    }
  } else {
    for (let k in oldAttrs) {
      if (!(k in attrs)) {
        removeAttribute.call(this, k);
      }
    }
    for (let k in attrs) {
      const val = attrs[k];
      if (val !== oldAttrs[k]) {
        setAttribute.call(this, k, val);
      }
    }
  }
}

function toClassObj(expr: string | number | { [c: string]: any }) {
  const result: { [c: string]: any } = {};
  switch (typeof expr) {
    case "string":
      // we transform here a list of classes into an object:
      //  'hey you' becomes {hey: true, you: true}
      const str = trim.call(expr);
      if (!str) {
        return {};
      }
      let words = split.call(str, wordRegexp);
      for (let i = 0, l = words.length; i < l; i++) {
        result[words[i]] = true;
      }
      return result;
    case "object":
      // this is already an object but we may need to split keys:
      // {'a': true, 'b c': true} should become {a: true, b: true, c: true}
      for (let key in expr as any) {
        const value = (expr as any)[key];
        if (value) {
          const words = split.call(key, wordRegexp);
          for (let word of words) {
            result[word] = value;
          }
        }
      }
      return result;

    case "undefined":
      return {};
    case "number":
      return { [expr as number]: true };
    default:
      return { [expr as any]: true };
  }
}

export function setClass(this: HTMLElement, val: any) {
  val = val === "" ? {} : toClassObj(val);
  // add classes
  const cl = this.classList;
  for (let c in val) {
    tokenListAdd.call(cl, c);
  }
}

export function updateClass(this: HTMLElement, val: any, oldVal: any) {
  oldVal = oldVal === "" ? {} : toClassObj(oldVal);
  val = val === "" ? {} : toClassObj(val);
  const cl = this.classList;
  // remove classes
  for (let c in oldVal) {
    if (!(c in val)) {
      tokenListRemove.call(cl, c);
    }
  }
  // add classes
  for (let c in val) {
    if (!(c in oldVal)) {
      tokenListAdd.call(cl, c);
    }
  }
}

export function makePropSetter(name: string): Setter<HTMLElement> {
  return function setProp(this: HTMLElement, value: any) {
    (this as any)[name] = value;
  };
}

export function isProp(tag: string, key: string): boolean {
  switch (tag) {
    case "input":
      return (
        key === "checked" ||
        key === "indeterminate" ||
        key === "value" ||
        key === "readonly" ||
        key === "disabled"
      );
    case "option":
      return key === "selected" || key === "disabled";
    case "textarea":
      return key === "value" || key === "readonly" || key === "disabled";
    case "select":
      return key === "value" || key === "disabled";
    case "button":
    case "optgroup":
      return key === "disabled";
  }
  return false;
}
