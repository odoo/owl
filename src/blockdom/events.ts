import { config } from "./config";

type EventHandlerSetter = (this: HTMLElement, data: any) => void;

interface EventHandlerCreator {
  setup: EventHandlerSetter;
  update: EventHandlerSetter;
}

export function createEventHandler(rawEvent: string): EventHandlerCreator {
  const eventName = rawEvent.split(".")[0];
  const capture = rawEvent.includes(".capture");
  if (rawEvent.includes(".synthetic")) {
    return createSyntheticHandler(eventName, capture);
  } else {
    return createElementHandler(eventName, capture);
  }
}

// Native listener
let nextNativeEventId = 1;
function createElementHandler(evName: string, capture: boolean = false): EventHandlerCreator {
  let eventKey = `__event__${evName}_${nextNativeEventId++}`;
  if (capture) {
    eventKey = `${eventKey}_capture`;
  }

  function listener(ev: Event) {
    const currentTarget = ev.currentTarget;
    if (!currentTarget || !document.contains(currentTarget as HTMLElement)) return;
    const data = (currentTarget as any)[eventKey];
    if (!data) return;
    config.mainEventHandler(data, ev, currentTarget);
  }

  function setup(this: HTMLElement, data: any) {
    (this as any)[eventKey] = data;
    this.addEventListener(evName, listener, { capture });
  }

  function update(this: HTMLElement, data: any) {
    (this as any)[eventKey] = data;
  }

  return { setup, update };
}

// Synthetic handler: a form of event delegation that allows placing only one
// listener per event type.
let nextSyntheticEventId = 1;
function createSyntheticHandler(evName: string, capture: boolean = false): EventHandlerCreator {
  let eventKey = `__event__synthetic_${evName}`;
  if (capture) {
    eventKey = `${eventKey}_capture`;
  }
  setupSyntheticEvent(evName, eventKey, capture);
  const currentId = nextSyntheticEventId++;
  function setup(this: HTMLElement, data: any) {
    const _data = (this as any)[eventKey] || {};
    _data[currentId] = data;
    (this as any)[eventKey] = _data;
  }
  return { setup, update: setup };
}

function nativeToSyntheticEvent(eventKey: string, event: Event) {
  let dom = event.target;
  while (dom !== null) {
    const _data = (dom as any)[eventKey];
    if (_data) {
      for (const data of Object.values(_data)) {
        const stopped = config.mainEventHandler(data, event, dom);
        if (stopped) return;
      }
    }
    dom = (dom as any).parentNode;
  }
}

const CONFIGURED_SYNTHETIC_EVENTS: { [event: string]: boolean } = {};

function setupSyntheticEvent(evName: string, eventKey: string, capture: boolean = false) {
  if (CONFIGURED_SYNTHETIC_EVENTS[eventKey]) {
    return;
  }
  document.addEventListener(evName, (event) => nativeToSyntheticEvent(eventKey, event), {
    capture,
  });
  CONFIGURED_SYNTHETIC_EVENTS[eventKey] = true;
}
