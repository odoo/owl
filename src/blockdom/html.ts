import type { VNode } from "./index";

const nodeProto = Node.prototype;

const nodeInsertBefore = nodeProto.insertBefore;
const nodeRemoveChild = nodeProto.removeChild;

class VHtml {
  html: string;
  parentEl?: HTMLElement | undefined;
  content: ChildNode[] = [];

  constructor(html: string) {
    this.html = html;
  }

  mount(parent: HTMLElement, afterNode: Node | null) {
    this.parentEl = parent;
    const template = document.createElement("template");
    template.innerHTML = this.html;
    this.content = [...(template.content.childNodes as any)];
    for (let elem of this.content) {
      nodeInsertBefore.call(parent, elem, afterNode);
    }
    if (!this.content.length) {
      const textNode = document.createTextNode("");
      this.content.push(textNode);
      nodeInsertBefore.call(parent, textNode, afterNode);
    }
  }

  moveBefore(other: VHtml | null, afterNode: Node | null) {
    const target = other ? other.content[0] : afterNode;
    const parent = this.parentEl;
    for (let elem of this.content) {
      nodeInsertBefore.call(parent, elem, target);
    }
  }

  patch(other: VHtml) {
    if (this === other) {
      return;
    }
    const html2 = other.html;
    if (this.html !== html2) {
      const parent = this.parentEl;
      // insert new html in front of current
      const afterNode = this.content[0];
      const template = document.createElement("template");
      template.innerHTML = html2;
      const content = [...(template.content.childNodes as any)];
      for (let elem of content) {
        nodeInsertBefore.call(parent, elem, afterNode);
      }
      if (!content.length) {
        const textNode = document.createTextNode("");
        content.push(textNode);
        nodeInsertBefore.call(parent, textNode, afterNode);
      }

      // remove current content
      this.remove();
      this.content = content;
      this.html = other.html;
    }
  }

  beforeRemove() {}

  remove() {
    const parent = this.parentEl;
    for (let elem of this.content) {
      nodeRemoveChild.call(parent, elem);
    }
  }

  firstNode(): Node {
    return this.content[0]!;
  }

  toString() {
    return this.html;
  }
}

export function html(str: string): VNode<VHtml> {
  return new VHtml(str);
}
