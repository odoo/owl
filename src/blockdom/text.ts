import type { VNode } from "./index";

const getDescriptor = (o: any, p: any) => Object.getOwnPropertyDescriptor(o, p)!;
const nodeProto = Node.prototype;
const characterDataProto = CharacterData.prototype;

const nodeInsertBefore = nodeProto.insertBefore;
const characterDataSetData = getDescriptor(characterDataProto, "data").set!;
const nodeRemoveChild = nodeProto.removeChild;

abstract class VSimpleNode {
  text: string;
  parentEl?: HTMLElement | undefined;
  el?: any;

  constructor(text: string) {
    this.text = text;
  }

  mountNode(node: Node, parent: HTMLElement, afterNode: Node | null) {
    this.parentEl = parent;
    nodeInsertBefore.call(parent, node, afterNode);
    this.el = node;
  }

  moveBefore(other: VText | null, afterNode: Node | null) {
    const target = other ? other.el! : afterNode;
    nodeInsertBefore.call(this.parentEl, this.el!, target);
  }

  beforeRemove() {}

  remove() {
    nodeRemoveChild.call(this.parentEl, this.el!);
  }

  firstNode(): Node {
    return this.el!;
  }

  toString() {
    return this.text;
  }
}

class VText extends VSimpleNode {
  mount(parent: HTMLElement, afterNode: Node | null) {
    this.mountNode(document.createTextNode(toText(this.text)), parent, afterNode);
  }

  patch(other: VText) {
    const text2 = other.text;
    if (this.text !== text2) {
      characterDataSetData.call(this.el!, toText(text2));
      this.text = text2;
    }
  }
}

class VComment extends VSimpleNode {
  mount(parent: HTMLElement, afterNode: Node | null) {
    this.mountNode(document.createComment(toText(this.text)), parent, afterNode);
  }

  patch() {}
}

export function text(str: string): VNode<VText> {
  return new VText(str);
}

export function comment(str: string): VNode<VComment> {
  return new VComment(str);
}

export function toText(value: any): string {
  switch (typeof value) {
    case "string":
      return value;
    case "number":
      return String(value);
    case "boolean":
      return value ? "true" : "false";
    default:
      return value || "";
  }
}
