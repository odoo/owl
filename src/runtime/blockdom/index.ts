export { config } from "./config";

export { toggler } from "./toggler";
export { createBlock } from "./block_compiler";
export { list } from "./list";
export { multi } from "./multi";
export { text, comment } from "./text";
export { html } from "./html";
export { createCatcher } from "./event_catcher";

export interface VNode<T = any> {
  mount(parent: HTMLElement, afterNode: Node | null): void;
  moveBeforeDOMNode(node: Node | null, parent?: HTMLElement): void;
  moveBeforeVNode(other: T | null, afterNode: Node | null): void;
  patch(other: T, withBeforeRemove: boolean): void;
  beforeRemove(): void;
  remove(): void;
  firstNode(): Node | undefined;

  el?: undefined | HTMLElement | Text;
  parentEl?: undefined | HTMLElement;
  isOnlyChild?: boolean | undefined;
  key?: any;
}

export type BDom = VNode<any>;

export function mount(vnode: VNode, fixture: HTMLElement, afterNode: Node | null = null) {
  vnode.mount(fixture, afterNode);
}

export function patch(vnode1: VNode, vnode2: VNode, withBeforeRemove: boolean = false) {
  vnode1.patch(vnode2, withBeforeRemove);
}

export function remove(vnode: VNode, withBeforeRemove: boolean = false) {
  if (withBeforeRemove) {
    vnode.beforeRemove();
  }
  vnode.remove();
}

export function withKey(vnode: VNode, key: any) {
  vnode.key = key;
  return vnode;
}
