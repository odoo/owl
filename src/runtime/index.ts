import { App } from "./app";
import {
  config,
  createBlock,
  html,
  list,
  mount as blockMount,
  multi,
  patch,
  remove,
  text,
  toggler,
  comment,
} from "./blockdom";
import { mainEventHandler } from "./event_handling";

config.shouldNormalizeDom = false;
config.mainEventHandler = mainEventHandler;

export const blockDom = {
  config,
  // bdom entry points
  mount: blockMount,
  patch,
  remove,
  // bdom block types
  list,
  multi,
  text,
  toggler,
  createBlock,
  html,
  comment,
};

export { App, mount } from "./app";
export { xml } from "./template_set";
export { Component } from "./component";
export type { ComponentConstructor } from "./component";
export { useComponent, useState } from "./component_node";
export { status } from "./status";
export { reactive, markRaw, toRaw } from "./reactivity";
export { useEffect, useEnv, useExternalListener, useRef, useChildSubEnv, useSubEnv } from "./hooks";
export { EventBus, whenReady, loadFile, markup } from "./utils";
export {
  onWillStart,
  onMounted,
  onWillUnmount,
  onWillUpdateProps,
  onWillPatch,
  onPatched,
  onWillRender,
  onRendered,
  onWillDestroy,
  onError,
} from "./lifecycle_hooks";
export { validate } from "./validation";
export { OwlError } from "./error_handling";

export const __info__ = {
  version: App.version,
};
