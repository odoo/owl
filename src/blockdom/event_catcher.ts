import { createEventHandler } from "./events";
import type { VNode } from "./index";

type EventsSpec = { [name: string]: number };

type Catcher = (child: VNode, handlers: any[]) => VNode;

export function createCatcher(eventsSpec: EventsSpec): Catcher {
  let setupFns: any[] = [];
  let removeFns: any[] = [];
  for (let name in eventsSpec) {
    let index = eventsSpec[name];
    let { setup, remove } = createEventHandler(name);
    setupFns[index] = setup;
    removeFns[index] = remove;
  }
  let n = setupFns.length;

  class VCatcher {
    child: VNode;
    handlers: any[];

    parentEl?: HTMLElement | undefined;
    afterNode: Node | null = null;

    constructor(child: VNode, handlers: any[]) {
      this.child = child;
      this.handlers = handlers;
    }

    mount(parent: HTMLElement, afterNode: Node | null) {
      this.parentEl = parent;
      this.afterNode = afterNode;
      this.child.mount(parent, afterNode);
      for (let i = 0; i < n; i++) {
        let origFn = this.handlers[i][0];
        const self = this;
        this.handlers[i][0] = function (ev: any) {
          const target = ev.target;
          let currentNode: any = self.child.firstNode();
          const afterNode = self.afterNode;
          while (currentNode !== afterNode) {
            if (currentNode.contains(target)) {
              return origFn.call(this, ev);
            }
            currentNode = currentNode.nextSibling;
          }
        };
        setupFns[i].call(parent, this.handlers[i]);
      }
    }

    moveBefore(other: VCatcher | null, afterNode: Node | null) {
      this.afterNode = null;
      this.child.moveBefore(other ? other.child : null, afterNode);
    }

    patch(other: VCatcher, withBeforeRemove: boolean) {
      if (this === other) {
        return;
      }
      this.handlers = other.handlers;
      this.child.patch(other.child, withBeforeRemove);
    }

    beforeRemove() {
      this.child.beforeRemove();
    }

    remove() {
      for (let i = 0; i < n; i++) {
        removeFns[i].call(this.parentEl!);
      }
      this.child.remove();
    }

    firstNode(): Node | undefined {
      return this.child.firstNode();
    }

    toString(): string {
      return this.child.toString();
    }
  }

  return function (child: VNode, handlers: any[]): VNode<VCatcher> {
    return new VCatcher(child, handlers);
  };
}
