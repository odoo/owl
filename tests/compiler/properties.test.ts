import { mount, patch } from "../../src/runtime/blockdom";
import { makeTestFixture, renderToBdom, renderToString, snapshotEverything } from "../helpers";

snapshotEverything();
test("changing an attribute with t-att-", () => {
  // render input with initial value
  const template = `<div t-att-value="v"/>`;
  const bnode1 = renderToBdom(template, { v: "zucchini" });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);

  expect(fixture.innerHTML).toBe('<div value="zucchini"></div>');

  const bnode2 = renderToBdom(template, { v: "potato" });
  patch(bnode1, bnode2);
  expect(fixture.innerHTML).toBe('<div value="potato"></div>');

  const bnode3 = renderToBdom(template, { v: "" });
  patch(bnode1, bnode3);
  // not sure about this. maybe we want to remove the attribute?
  expect(fixture.innerHTML).toBe('<div value=""></div>');
});

test("dynamic input value: falsy values", () => {
  // 0 doesn't fall back to empty string
  expect(renderToBdom(`<input t-att-value="0"/>`)).toEqual({ data: [new String("0")] });
  expect(renderToBdom(`<input t-att-value="false"/>`)).toEqual({ data: [new String("")] });
  expect(renderToBdom(`<input t-att-value="undefined"/>`)).toEqual({ data: [new String("")] });
  expect(renderToBdom(`<input t-att-value="''"/>`)).toEqual({ data: [new String("")] });
});

test("updating property with falsy value", async () => {
  // render input with initial value
  const template = `<input t-att-value="v"></input>`;
  const bnode1 = renderToBdom(template, { v: false });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);

  const input = fixture.querySelector("input")!;
  expect(input.value).toBe("");

  patch(bnode1, renderToBdom(template, { v: "owl" }));
  expect(input.value).toBe("owl");

  patch(bnode1, renderToBdom(template, { v: false }));
  expect(input.value).toBe("");

  patch(bnode1, renderToBdom(template, { v: "owl" }));
  expect(input.value).toBe("owl");

  patch(bnode1, renderToBdom(template, { v: undefined }));
  expect(input.value).toBe("");

  patch(bnode1, renderToBdom(template, { v: "owl" }));
  expect(input.value).toBe("owl");

  patch(bnode1, renderToBdom(template, { v: null }));
  expect(input.value).toBe("");
});


test("input type= checkbox, with t-att-checked", () => {
  const template = `<input type="checkbox" t-att-checked="flag"/>`;
  const result = renderToString(template, { flag: true });
  expect(result).toBe(`<input type="checkbox">`);
});

test("various boolean html attributes", () => {
  // will cause the template to be snapshotted
  renderToString(`
      <div>
        <input type="checkbox" checked="checked"/>
        <input checked="checked"/>
        <div checked="checked"/>
        <div selected="selected"/>
        <option selected="selected" other="1"/>
        <input readonly="readonly"/>
        <button disabled="disabled"/>
      </div>
      `);
});

test("input with t-att-value", () => {
  // render input with initial value
  const template = `<input  t-att-value="v"/>`;
  const bnode1 = renderToBdom(template, { v: "zucchini" });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const input = fixture.querySelector("input")!;
  expect(input.value).toBe("zucchini");

  // change value manually in input, to simulate user input
  input.value = "tomato";
  expect(input.value).toBe("tomato");

  // rerender with a different value, and patch actual dom, to check that
  // input value was properly reset by owl
  const bnode2 = renderToBdom(template, { v: "potato" });
  patch(bnode1, bnode2);
  expect(input.value).toBe("potato");
});

test("input with t-att-value (patching with same value", () => {
  // render input with initial value
  const template = `<input t-att-value="v"/>`;
  const bnode1 = renderToBdom(template, { v: "zucchini" });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const input = fixture.querySelector("input")!;
  expect(input.value).toBe("zucchini");

  // change value manually in input, to simulate user input
  input.value = "tomato";
  expect(input.value).toBe("tomato");

  const bnode2 = renderToBdom(template, { v: "zucchini" });
  patch(bnode1, bnode2);
  expect(input.value).toBe("zucchini");
});

test("input, type checkbox, with t-att-checked (patching with same value", () => {
  // render input with initial value
  const template = `<input type="checkbox" t-att-checked="v"/>`;
  const bnode1 = renderToBdom(template, { v: true });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const input = fixture.querySelector("input")!;
  expect(input.checked).toBe(true);

  // change checked manually in input, to simulate user input
  input.checked = false;
  expect(input.checked).toBe(false);

  const bnode2 = renderToBdom(template, { v: true });
  patch(bnode1, bnode2);
  expect(input.checked).toBe(true);
});

test("input of type checkbox with t-att-indeterminate", () => {
  const template = `<input type="checkbox" t-att-indeterminate="v"/>`;
  const bnode1 = renderToBdom(template, { v: true });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const input = fixture.querySelector("input")!;
  expect(input.indeterminate).toBe(true);
});

test("textarea with t-att-value", () => {
  // render textarea with initial value
  const template = `<textarea t-att-value="v"/>`;
  const bnode1 = renderToBdom(template, { v: "zucchini" });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const elm = fixture.querySelector("textarea")!;
  expect(elm.value).toBe("zucchini");

  // change value manually in textarea, to simulate user textarea
  elm.value = "tomato";
  expect(elm.value).toBe("tomato");

  // rerender with a different value, and patch actual dom, to check that
  // textarea value was properly reset by owl
  const bnode2 = renderToBdom(template, { v: "potato" });
  patch(bnode1, bnode2);
  expect(elm.value).toBe("potato");
});

test("select with t-att-value", () => {
  const template = `
      <select t-att-value="value">
        <option value="potato">Potato</option>
        <option value="tomato">Tomato</option>
        <option value="onion">Onion</option>
      </select>`;
  const bnode1 = renderToBdom(template, { value: "tomato" });
  const fixture = makeTestFixture();
  mount(bnode1, fixture);
  const elm = fixture.querySelector("select")!;
  expect(elm.value).toBe("tomato");

  elm.value = "potato";
  expect(elm.value).toBe("potato");

  // rerender with a different value, and patch actual dom, to check that
  // select value was properly reset by owl
  const bnode2 = renderToBdom(template, { value: "onion" });
  patch(bnode1, bnode2);
  expect(elm.value).toBe("onion");
});
