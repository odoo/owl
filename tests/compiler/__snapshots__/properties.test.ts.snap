// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changing an attribute with t-att- 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<div block-attribute-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let attr1 = ctx['v'];
    return block1([attr1]);
  }
}"
`;

exports[`dynamic input value: falsy values 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((0) === 0 ? 0 : ((0) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`dynamic input value: falsy values 2`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((false) === 0 ? 0 : ((false) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`dynamic input value: falsy values 3`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((undefined) === 0 ? 0 : ((undefined) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`dynamic input value: falsy values 4`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String(('') === 0 ? 0 : (('') || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`input of type checkbox with t-att-indeterminate 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input type=\\"checkbox\\" block-property-0=\\"indeterminate\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new Boolean(ctx['v']);
    return block1([prop1]);
  }
}"
`;

exports[`input type= checkbox, with t-att-checked 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input type=\\"checkbox\\" block-property-0=\\"checked\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new Boolean(ctx['flag']);
    return block1([prop1]);
  }
}"
`;

exports[`input with t-att-value (patching with same value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((ctx['v']) === 0 ? 0 : ((ctx['v']) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`input with t-att-value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((ctx['v']) === 0 ? 0 : ((ctx['v']) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`input, type checkbox, with t-att-checked (patching with same value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input type=\\"checkbox\\" block-property-0=\\"checked\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new Boolean(ctx['v']);
    return block1([prop1]);
  }
}"
`;

exports[`select with t-att-value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<select block-property-0=\\"value\\"><option value=\\"potato\\">Potato</option><option value=\\"tomato\\">Tomato</option><option value=\\"onion\\">Onion</option></select>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((ctx['value']) === 0 ? 0 : ((ctx['value']) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`textarea with t-att-value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<textarea block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((ctx['v']) === 0 ? 0 : ((ctx['v']) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`updating property with falsy value 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<input block-property-0=\\"value\\"/>\`);
  
  return function template(ctx, node, key = \\"\\") {
    let prop1 = new String((ctx['v']) === 0 ? 0 : ((ctx['v']) || \\"\\"));
    return block1([prop1]);
  }
}"
`;

exports[`various boolean html attributes 1`] = `
"function anonymous(app, bdom, helpers
) {
  let { text, createBlock, list, multi, html, toggler, comment } = bdom;
  
  let block1 = createBlock(\`<div><input type=\\"checkbox\\" checked=\\"checked\\"/><input checked=\\"checked\\"/><div checked=\\"checked\\"/><div selected=\\"selected\\"/><option selected=\\"selected\\" other=\\"1\\"/><input readonly=\\"readonly\\"/><button disabled=\\"disabled\\"/></div>\`);
  
  return function template(ctx, node, key = \\"\\") {
    return block1();
  }
}"
`;
